### PR TITLE
Fix child processes not being reaped when `Process.detach` used

### LINKED ISSUE #3313: Puma cluster not reaping child processes when PID is 1 with Puma 6.4.1
**Describe the bug**

We have a separate fleet of Puma workers to handle ActionCable, and since upgrading to 6.4.1 we have seen a significant increase of unhealthy pods and `can't alloc thread` error messages. In addition, the pod's readiness checks started to fail, causing Kubernetes to periodically shutdown and restart the Puma container.

We rolled back to 6.4.0 and saw a dramatic drop in errors. See https://gitlab.com/gitlab-com/gl-infra/production/-/issues/17372 for the full details.

**Puma config:**

<details>

```
# frozen_string_literal: true

# Load "path" as a rackup file.
#
# The default is "config.ru".
#
rackup '/srv/gitlab/config.ru'
pidfile "#{ENV['HOME']}/puma.pid"
state_path "#{ENV['HOME']}/puma.state"

stdout_redirect '/srv/gitlab/log/puma.stdout.log',
  '/srv/gitlab/log/puma.stderr.log',
  true

# Configure "min" to be the minimum number of threads to use to answer
# requests and "max" the maximum.
#
# The default is "0, 16".
#
threads (ENV['PUMA_THREADS_MIN'] ||= '1').to_i , (ENV['PUMA_THREADS_MAX'] ||= '16').to_i

# By default, workers accept all requests and queue them to pass to handlers.
# When false, workers accept the number of simultaneous requests configured.
#
# Queueing requests generally improves performance, but can cause deadlocks if
# the app is waiting on a request to itself. See https://github.com/puma/puma/issues/612
#
# When set to false this may require a reverse proxy to handle slow clients and
# queue requests before they reach puma. This is due to disabling HTTP keepalive
queue_requests false

# Bind the server to "url". "tcp://", "unix://" and "ssl://" are the only
# accepted protocols.

# We want to provide the ability to enable individually control HTTP (`INTERNAL_PORT`)
# HTTPS (`SSL_INTERNAL_PORT`):
#
# 1. HTTP on, HTTPS on: Since `INTERNAL_PORT` is configured, we listen on it.
# 2. HTTP on, HTTPS off: If we don't specify either port, we default to HTTP
#    because SSL requires a certificate and key to work.
# 3. HTTP off, HTTPS on: `SSL_INTERNAL_PORT` is enabled but
#   `INTERNAL_PORT` is not set.
http_port = ENV['INTERNAL_PORT'] || '8080'
http_addr =
  if ENV['INTERNAL_PORT'] || (!ENV['INTERNAL_PORT'] && !ENV['SSL_INTERNAL_PORT'])
    "0.0.0.0"
  else
    # If HTTP is disabled, we still need to listen to 127.0.0.1 for health checks.
    "127.0.0.1"
  end

bind "tcp://#{http_addr}:#{http_port}"

if ENV['SSL_INTERNAL_PORT']
  ssl_params = {
    cert: ENV['PUMA_SSL_CERT'],
    key: ENV['PUMA_SSL_KEY'],
  }

  ssl_params[:ca] = ENV['PUMA_SSL_CLIENT_CERT'] if ENV['PUMA_SSL_CLIENT_CERT']
  ssl_params[:key_password_command] = ENV['PUMA_SSL_KEY_PASSWORD_COMMAND'] if ENV['PUMA_SSL_KEY_PASSWORD_COMMAND']
  ssl_params[:ssl_cipher_filter] = ENV['PUMA_SSL_CIPHER_FILTER'] if ENV['PUMA_SSL_CIPHER_FILTER']
  ssl_params[:verify_mode] = ENV['PUMA_SSL_VERIFY_MODE'] || 'none'

  ssl_bind '0.0.0.0', ENV['SSL_INTERNAL_PORT'], ssl_params
end

workers (ENV['WORKER_PROCESSES'] ||= '3').to_i

require "/srv/gitlab/lib/gitlab/cluster/lifecycle_events"

on_restart do
  # Signal application hooks that we're about to restart
  Gitlab::Cluster::LifecycleEvents.do_before_master_restart
end

before_fork do
  # Signal application hooks that we're about to fork
  Gitlab::Cluster::LifecycleEvents.do_before_fork
end

Gitlab::Cluster::LifecycleEvents.set_puma_options @config.options
on_worker_boot do
  # Signal application hooks of worker start
  Gitlab::Cluster::LifecycleEvents.do_worker_start
end

on_worker_shutdown do
  # Signal application hooks that a worker is shutting down
  Gitlab::Cluster::LifecycleEvents.do_worker_stop
end

# Preload the application before starting the workers; this conflicts with
# phased restart feature. (off by default)
preload_app!

tag 'gitlab-puma-worker'

# Verifies that all workers have checked in to the master process within
# the given timeout. If not the worker process will be restarted. Default
# value is 60 seconds.
#
worker_timeout (ENV['WORKER_TIMEOUT'] ||= '60').to_i

# https://github.com/puma/puma/blob/master/5.0-Upgrade.md#lower-latency-better-throughput
wait_for_less_busy_worker (ENV['PUMA_WAIT_FOR_LESS_BUSY_WORKER'] ||= '0.001').to_f

# Use json formatter
require "/srv/gitlab/lib/gitlab/puma_logging/json_formatter"

json_formatter = Gitlab::PumaLogging::JSONFormatter.new
log_formatter do |str|
  json_formatter.call(str)
end

require "/srv/gitlab/lib/gitlab/puma/error_handler"

error_handler = Gitlab::Puma::ErrorHandler.new(ENV['RAILS_ENV'] == 'production')

lowlevel_error_handler do |ex, env, status_code|
  error_handler.execute(ex, env, status_code)
end
```

</details>

Please copy-paste your Puma config AND your command line options here.

```shell
/srv/gitlab/bin/bundle exec puma --environment production --config /srv/gitlab/config/puma.rb /srv/gitlab/config.ru
```

Example process list:

```
$ ps -ef
UID          PID    PPID  C STIME TTY          TIME CMD
git            1       0  0 20:07 ?        00:00:34 puma 6.4.1 (tcp://0.0.0.0:8080) [gitlab-puma-worker]
git           41       1  0 20:07 ?        00:00:21 /usr/local/bin/gitlab-logger /var/log/gitlab
git           60       1  0 20:07 ?        00:00:47 ruby /srv/gitlab/bin/metrics-server
git           63       1  1 20:07 ?        00:00:54 puma: cluster worker 0: 1 [gitlab-puma-worker]
git           65       1  1 20:07 ?        00:00:55 puma: cluster worker 1: 1 [gitlab-puma-worker]
git           67       1  1 20:07 ?        00:00:59 puma: cluster worker 2: 1 [gitlab-puma-worker]
git           69       1  1 20:07 ?        00:00:55 puma: cluster worker 3: 1 [gitlab-puma-worker]
git          546       0  0 21:37 pts/0    00:00:00 bash
git          556     546  0 21:37 pts/0    00:00:00 ps -ef
```

Note we are running Puma as PID 1. I don't believe `--fork-worker` is being used.

**To Reproduce**

I'm working on a reproduction step right now. I suspect https://github.com/puma/puma/pull/3255 might have caused this issue. I didn't see `reaped unknown child process` messages for this ActionCable fleet, though I did see it in other fleet of workers that didn't appear to have increased error rates.

One thing I observed is that previously `wait_workers` could call `Process.kill(0, w.pid)` to verify that each worker was still running: https://github.com/puma/puma/blob/52eff8d3344fa0a4daa8d00ba1de74cb746a5ff4/lib/puma/cluster.rb#L515-L523

Now that check only happens if `fork_worker` is enabled? https://github.com/puma/puma/blob/a2870252e3c525f6529358807faee1169f28270e/lib/puma/cluster.rb#L565-L573

@casperisfine @nateberkopec What do you think about reverting https://github.com/puma/puma/pull/3255 or putting it behind some configuration parameter?

**Expected behavior**

No errors.

- Linux x86 running Puma 6.4.1 inside Debian bookworm container
- Google Kubernetes Engine

### PR DESCRIPTION
**Pull request by stanhu**
### Description

Starting with Puma v6.4.1, we observed that killed Puma cluster workers were never being restarted when the parent was run as PID 1. For example, I issued a `kill 44` and PID 44 remained in the `defunct` state:

```
git@gitlab-webservice-default-78664bb757-2nxvh:/var/log/gitlab$ ps -ef
UID          PID    PPID  C STIME TTY          TIME CMD
git            1       0  0 Jan09 ?        00:01:39 puma 6.4.1 (tcp://0.0.0.0:8080) [gitlab-puma-worker]
git           23       1  0 Jan09 ?        00:05:46 /usr/local/bin/gitlab-logger /var/log/gitlab
git           41       1  0 Jan09 ?        00:01:55 ruby /srv/gitlab/bin/metrics-server
git           44       1  0 Jan09 ?        00:02:41 [ruby] <defunct>
git           46       1  0 Jan09 ?        00:02:38 puma: cluster worker 1: 1 [gitlab-puma-worker]
git           48       1  0 Jan09 ?        00:02:42 puma: cluster worker 2: 1 [gitlab-puma-worker]
git           49       1  0 Jan09 ?        00:02:41 puma: cluster worker 3: 1 [gitlab-puma-worker]
git         5205       0  0 21:57 pts/0    00:00:00 bash
git         5331    5205  0 22:00 pts/0    00:00:00 ps -ef
```

Further investigation showed that the introduction of `Process.wait2(-1, Process::WNOHANG)` in #3255 never appears to return anything when `Process.detach` is run on some process that has not exited. This bug appears to be present from Ruby 2.6 to 3.2, but has been been fixed in Ruby 3.3: https://bugs.ruby-lang.org/issues/19837

Previously `Process.wait(w.pid, Process::WNOHANG)` was called on each known worker PID. #3255 changed this behavior to do this only if the `fork_worker` config parameter were enabled, but it seems that we should always do this to ensure that terminated workers are reaped in a timely manner.

Closes #3313 

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

**Comment by MSP-Greg** [comment]
@stanhu

Thanks for the PR.  LGTM, but I'd like to run some additional tests.  Currently knee deep in some dumb s*&(...

**Comment by dentarg** [comment]
Does your latest comment https://github.com/puma/puma/issues/3313#issuecomment-1886629196 change anything for these changes @stanhu?

**Comment by dentarg** [comment]
Also, from your investigation in https://github.com/puma/puma/issues/3313, it looks like it wouldn't be impossible to add some test (using Docker) for this PR and https://github.com/puma/puma/pull/3255? Could be something only GitHub Actions run, not part of `bundle exec rake`. What do you and others (@byroot) think about that? Is it worth it?

**Comment by stanhu** [comment]
@dentarg I'm still investigating how it's possible for the application to interfere with `Process.wait2(-1, Process::WNOHANG)`, presumably by trapping certain signals. 

I think this PR can't hurt, and it restores the previous behavior of checking known PIDs in addition to checking child processes.

**Comment by byroot** [comment]
> Could be something only GitHub Actions run, not part of `bundle exec rake`. What do you and others (@byroot) think about that? Is it worth it?

I think it would be worth it. Note that an alternative to running as PID 1 is to declare the process as a SUBREAPER: https://github.com/Shopify/pitchfork/blob/50f3e3389218e6e82c65638ab3c91f805ec02c4b/ext/pitchfork_http/child_subreaper.h

It's a Linux only API, but would allow to do it as part of the main test suite when ran on Linux. I made a gem a while ago you could use: https://rubygems.org/gems/child_subreaper

**Comment by stanhu** [comment]
I've updated this pull request to reflect that it appears that this works around a Ruby 3.1/3.2 bug with `Process.detach`: https://github.com/puma/puma/issues/3313#issuecomment-1887579232

I think a test is a good idea, but that will take me longer to get to at the moment.

**Comment by byroot** [comment]
> a Ruby 3.1/3.2 bug with `Process.detach`

Would be worth reporting upstream. Even if it's fixed in 3.3, it may be worth a backport.

**Comment by stanhu** [comment]
Good idea. I filed this bug: https://bugs.ruby-lang.org/issues/20181

**Comment by stanhu** [comment]
@MSP-Greg @dentarg I've added an integration test that appears to reproduce https://bugs.ruby-lang.org/issues/20181. This test fails in `master`, but passes in this branch.

**Comment by stanhu** [comment]
It seems this bug was already reported in https://bugs.ruby-lang.org/issues/19837 and fixed in the `ruby_3_2` and `ruby_3_1` branches, but the fixes aren't in a release yet.

**Comment by stanhu** [comment]
@MSP-Greg Is there anything else I can help with to get this merged?

**Comment by stanhu** [comment]
@MSP-Greg Sorry to bother you again, but would you have a moment to review?

**Comment by stanhu** [comment]
@MSP-Greg I did not see the comment. Is it published?

**Comment by MSP-Greg** [comment]
> I did not see the comment. Is it published?

Sorry, my mistake.  I didn't click 'Publish Review'.  Until one does that, the reviewer can see it, but no one else...

Do you see it now?

File: test/test_integration_cluster.rb [hunk 1]
```diff
@@ -175,6 +175,35 @@ def test_stuck_external_term_spawn
...
+  # https://bugs.ruby-lang.org/issues/19837. However,
+  # `Process.wait2(<child pid>)` still works properly. This bug has
+  # been fixed in Ruby 3.3.
+  def test_workers_respawn_with_process_detach
+    skip_unless_signal_exist? :KILL
+
+    config = 'test/config/process_detach_before_fork.rb'
+
+    worker_respawn(0, workers, config) do |phase0_worker_pids|
+      last = phase0_worker_pids.last
```
[the remark this answers is not in the document]
Comment (stanhu) [reply test/test_integration_cluster.rb]: Good call, we do not need it. Dropped.

**Comment by MSP-Greg** [comment]
@stanhu

Thank you for the PR.  Sorry for the delay.

**Comment by stanhu** [comment]
@nateberkopec Would you mind releasing an update with this? We're blocked on Puma 6.4.0 until this gets shipped.

**Comment by stanhu** [comment]
@nateberkopec Sorry to bother you again. Could you find some time to release a new version of Puma?

**Comment by stanhu** [comment]
FYI, this pull request is still needed in Ruby 3.1 and 3.2 because `Process.waitpid2(-1, Process::WNOHANG)` does not work properly in Ruby 3.1.5 and 3.2.4. See https://bugs.ruby-lang.org/issues/20490?next_issue_id=20489#note-1.

**Comment by byroot** [comment]
Even if it was fixed in tiny releases, I'd still recommend keeping the workaround until the entire line is no longer supported by puma.

**Comment by stanhu** [comment]
I agree. I'm just trying to raise the need for a release because any application that launches a subprocess with Puma will find that their cluster workers no longer will be reaped from Puma 6.4.1 to 6.4.2.

**Pull request merged**